core/src/main/scala/com/pingcap/tispark/TiBatchWrite.scala

```diff
@@ -304,15 +299,41 @@ class TiBatchWrite(@transient val df: DataFrame,
       }
     }
 
-    val deduplicatedTiRowRdd = deduplicateIfNecessary(tiRowRdd, tiTableInfo.isPkHandle)
-
     // get timestamp as start_ts
     val startTimeStamp = tiSession.getTimestamp
 
-    // TODO support insert ignore
-    val toBeCheckedRdd = getKeysNeedCheck(deduplicatedTiRowRdd, startTimeStamp)
-    checkConflictWithInsertKeys(toBeCheckedRdd)
-    val encodedTiRowRDD = generateRDDToBeInserted(toBeCheckedRdd)
+    // for partition table, we need calculate each row and tell which physical table
+    // that row is belong to.
+    // currently we only support upsert and insert.
+    val constraintCheckIsNeeded = handleCol != null || uniqueIndices.nonEmpty
+
+    val encodedTiRowRDD = if (constraintCheckIsNeeded) {
+      val wrappedRowRdd = if (tiTableInfo.isPkHandle) {
+        tiRowRdd.map { row =>
+          WrappedRow(row, extractHandleId(row))
+        }
+      } else {
+        val start = getAutoTableIdStart(tiRowRdd.count)
+        tiRowRdd.zipWithIndex.map { data =>
+          WrappedRow(data._1, data._2 + start)
+        }
+      }
+
+      val distinctWrappedRowRdd = deduplicate(wrappedRowRdd)
+
+      val rddDeletion = generateDataToBeRemovedRdd(distinctWrappedRowRdd, startTimeStamp)
+      if (!options.replace && !rddDeletion.isEmpty()) {
+        throw new TiBatchWriteException("data to be inserted has conflicts with TiKV data")
+      }
+      generateKV(distinctWrappedRowRdd, remove = false) ++ generateKV(rddDeletion, remove = true)
```

Comment:
For duplicated data, do we need to delete the old version first, then insert the new version? If yes, then how can we ensure that at runtime?

Reply:
Assume a key need to be written 2 times in this batchwrite(delete and insert), during 2pc, what happens if one of the write is chosen as the primary key(or more inopportune, what happens if the insert action of the key is chosen as the primary key during pre-write)?

Reply:
The same key will not be written twice in one batch.

Reply:
A new test case is added to cover this. It seems ok to merge two rdd.

Reply:
Resolved.